parsec/core/sync_monitor.py

```diff
@@ -2,155 +2,412 @@
+            ctx = ctxs.get(id)
+        else:
+            ctx = ctxs.get(workspace_id)
+        if ctx.local_change(id):
+            early_wakeup.set()
+
+    def _on_realm_vlobs_updated(sender, realm_id, checkpoint, src_id, src_version):
+        try:
+            ctx = ctxs.get(realm_id)
+        except FSWorkspaceNotFoundError:
+            # If the workspace is not present, nothing to do
+            # (this can happen when a workspace is just shared with
+            # us, hence we receive vlob updated events before having
+            # added the workspace entry to our user manifest)
+            return
+        if ctx.remote_change(src_id):
+            early_wakeup.set()
+
+    def _on_sharing_updated(sender, new_entry, previous_entry):
+        if new_entry.role is None:
+            # No longer access to the workspace
+            ctxs.discard(new_entry.id)
+        else:
             try:
-                await self.user_fs.sync()
-                user_manifest = self.user_fs.get_user_manifest()
-                for entry in user_manifest.workspaces:
-                    workspace = self.user_fs.get_workspace(entry.id)
-                    await workspace.sync("/")
-            finally:
-                self.event_bus.send("sync_monitor.reconnection_sync.done")
+                ctx = ctxs.get(new_entry.id)
+            except FSWorkspaceNotFoundError:
+                return
+            if ctx.role_updated(new_entry.role):
+                early_wakeup.set()
+
+    with event_bus.connect_in_context(
+        ("fs.entry.updated", _on_entry_updated),
+        ("backend.realm.vlobs_updated", _on_realm_vlobs_updated),
+        ("sharing.updated", _on_sharing_updated),
+    ):
+
+        wait_times = []
+        event_bus.send("sync_monitor.reconnection_sync.started")
+        wait_times.append(await ctxs.get(user_fs.user_manifest_id).tick())
+        user_manifest = user_fs.get_user_manifest()
+        for entry in user_manifest.workspaces:
+            if entry.role is not None:
+                try:
+                    wait_times.append(await ctxs.get(entry.id).tick())
+                except Exception as exc:
+                    logger.error("Sync monitor has crashed", workspace_id=entry.id, exc_info=exc)
+                    # Reset sync context which is now in an undefined state
+                    ctxs.discard(entry.id)
+                    ctx = ctxs.get(entry.id)
+                    # Add small cooldown just to be sure not end up in a crazy busy error loop
+                    ctx.due_time += 5
+        event_bus.send("sync_monitor.reconnection_sync.done")
+
+        event_bus.send("sync_monitor.ready")
+        while True:
+            with trio.move_on_at(min(wait_times)):
+                await early_wakeup.wait()
+                early_wakeup.clear()
+            wait_times.clear()
+            # Force a sleep to block here when time is frozen in tests
+            await trio.sleep(0.001)
```

Comment:
Again, is this really what we want?

Reply:
I guess so...

Another solution would be to define:
```python
async def monitor_wakeup():
    return
```

use it instead of `trio.sleep` and monkeypatch it in the tests
it's more cumbersome though, what do you prefer ?